app/src/main/java/com/graeberj/taskman/auth/presentation/login/LoginViewModel.kt

```diff
@@ -1,27 +1,81 @@
 package com.graeberj.taskman.auth.presentation.login
 
 import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
 import com.graeberj.taskman.auth.domain.repository.AuthRepository
+import com.graeberj.taskman.auth.domain.usecase.ValidateFormUseCase
+import com.graeberj.taskman.core.domain.preferences.Preferences
+import com.graeberj.taskman.util.Resource
 import dagger.hilt.android.lifecycle.HiltViewModel
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
 import javax.inject.Inject
 
 
 @HiltViewModel
 class LoginViewModel @Inject constructor(
-    private val repository: AuthRepository
+    private val repository: AuthRepository,
+    private val validateForm: ValidateFormUseCase,
+    private val preferences: Preferences
 ) : ViewModel() {
+
     private val _state = MutableStateFlow(LoginState())
     val state = _state.asStateFlow()
 
     fun onEvent(event: LoginEvent) {
         when (event) {
-            LoginEvent.Submit -> TODO()
-            LoginEvent.TogglePasswordVisibility -> TODO()
-            is LoginEvent.ValidateEmail -> TODO()
-            is LoginEvent.ValidatePassword -> TODO()
+            LoginEvent.Submit -> submit()
+            LoginEvent.TogglePasswordVisibility -> {
+                _state.update { it.copy( isPasswordHidden = !it.isPasswordHidden )}
+            }
+
+            is LoginEvent.ValidateEmail -> {
+                _state.update {
+                    it.copy(
+                        email = event.email,
+                        isEmailValid = validateForm.validateEmail(event.email)
+                    )
+                }
+            }
+
+            is LoginEvent.ValidatePassword -> {
+                _state.update {
+                    it.copy(
+                        password = event.password,
+                        isPasswordValid = validateForm.validatePassword(event.password)
+                    )
+                }
+            }
         }
 
     }
+
+    private fun submit() {
+        val currentState = _state.value
+        if (currentState.isEmailValid && currentState.isPasswordValid) {
+            viewModelScope.launch {
+                when (val result = repository.loginUser(currentState.email, currentState.password)) {
+                    is Resource.Success -> {
+                        result.data?.let { user ->
+                            preferences.saveToken(user.token)
+                            preferences.saveFullName(user.fullName)
+                            preferences.saveUserId(user.userId)
+                        }
+                        _state.value = currentState.copy(
+                            isUserLoggedIn = true
+                        )
```

Comment:
In general, I'd always use `_stateFlow.update { it.copy(...) }` to avoid race conditions. This is how it could happen:

*StateFlow holds value A*
- Thread 1 calls `copy()` on StateFlow with value A
- Before Thread 1 assigns its changed copy, Thread 2 updates StateFlow with value B
- Now, Thread 1 assigns its changed copy, which is still based on value A though, but the latest value is already value B. Therefore, value B would never be considered here.

Not always an issue, but I worked on highly stateful apps with frequently changing states, in which this really bit me, since such bugs are super hard to find.
Using `_state.value = _state.value.copy(...)` is fine if you always call this from a single thread. But in ViewModels, you're very likely to not do that as soon as you'd change the state inside a `viewModelScope`.